### PR TITLE
honor-first-shutdown

### PR DESCRIPTION
**Pull request by ejb42**
This pull request is will make the first shutdown request be the only shutdown request honored. The scenario I specifically ran into was a service configured as FailureAction=reboot. During a normal poweroff shutdown this service failed and changed the shutdown to a reboot type. This happened since systemd honored the last shutdown request received.
The two test files can be used to both verify the problem exists and verify the problem is fixed. This issue was seen on a fedora 31 machine using systemd v243.
Steps for the test files
- Install the service file and enable
- Install the shell script in /home (or any location - just remember to change the service file accordingly)
reboot your systemd image so the new service is started normally
perform a poweroff to see it change to a reboot
Install this fix and perform the same steps and see the poweroff complete normally

### CODE REVIEW

**Comment by keszybz** [comment]
This solution seems pretty intrusive. The original problem was that a FailureAction= from a failing unit during shutdown overrode the shutdown method. I agree that this seems counter-intuitive and unwanted. But I think the right solution here would be to ignore FailureAction= if a reboot is already in progress. I.e. instead of rejecting the job after it has been created, the creation of the job should be skipped.

Also, we already have ManagerObjective state, and this should be checked, instead of creating a new flag.

**Comment by keszybz** [comment]
Oh, and please remove all the temporary parts like `####` and commented out fields — we never want to merge stuff like that.

**Comment by ejb42** [comment]
keszybz - Thank you for your review. I will work on your suggestion asap.

**Comment by ejb42** [comment]
Zbigniew,

I am trying to address your comments but I have run into two points of confusion.

I am thinking the manager_add_job() function is the point where the job gets created, is that not correct?
I thought the logic I added allowed the first shutdown target job to be created but not any other starting type of job.
Please point me to where the SPECIAL_(reboot/poweroff/halt)_TARGET jobs are created,  I am having trouble finding that.

I was not aware of the ManagerObjective, however it does not appear to change when the shutdown target is active.
In my testing, I added a change to the objective value once the shutdown target was created, but the manager_loop
exits and the shutdown does not happen like it should - none of the services get shutdown. Is that really something I should
use for this purpose?

Thanks,
-Jay

**Comment by keszybz** [comment]
See `start_unit()` in systemctl.c. It just queues a halt.target/isolate job or similar.

**Comment by ejb42** [comment]
@keszybz  I believe I have addressed your rework concerns. I have tested the changes in the same fashion as documented. Please let me know if there are any other issues to address. Thanks.

File: src/test/test-honor-first-shutdown.sh [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+echo "This is the test script"
+count=1
+while [ $count -eq 1 ]; do
+    let count=count
```
Comment (keszybz) [inline src/test/test-honor-first-shutdown.sh]: This does an infinite busy loop. If this is intended, something more obvious that actually sleeps would be better:
```
while true; do
  sleep 1;
done
```

File: src/test/test-honor-first-shutdown.service [hunk 2]
```diff
@@ -0,0 +1,15 @@
+[Unit]
+Description=Test service to test systemd issues
+After=multi-user.target
+
+[Service]
+ExecStart=/home/test.sh
+ExecStop=/usr/bin/pkill -9 test.sh
```
Comment (keszybz) [inline src/test/test-honor-first-shutdown.service]: No absolute paths for binaries in the standard path.
Please use a signal name, in principle signal numbers vary by architecture.
This would pull in a dependency on procps-ng, which we don't want. Just use `kill -SIGKILL $MAINPID`.

File: src/test/test-honor-first-shutdown.service [hunk 3]
```diff
@@ -0,0 +1,15 @@
+[Unit]
+Description=Test service to test systemd issues
```
Comment (keszybz) [inline src/test/test-honor-first-shutdown.service]: That doesn't explain anything. It should be either be made more meaningful or dropped.

File: src/core/manager.h [hunk 4]
```diff
@@ -49,6 +49,7 @@ typedef enum ManagerObjective {
         MANAGER_HALT,
         MANAGER_KEXEC,
         MANAGER_SWITCH_ROOT,
+        MANAGER_SHUTDOWN,
```
Comment (keszybz) [inline src/core/manager.h]: "Shutdown" is either "halt" or "poweroff". I don't think there's any reason to add another enum here.

File: src/core/unit.c [hunk 5]
```diff
@@ -2600,8 +2600,12 @@ void unit_notify(Unit *u, UnitActiveState os, UnitActiveState ns, UnitNotifyFlag
...
 
                 if (os != UNIT_FAILED && ns == UNIT_FAILED) {
-                        reason = strjoina("unit ", u->id, " failed");
-                        emergency_action(m, u->failure_action, 0, u->reboot_arg, unit_failure_action_exit_status(u), reason);
+                        if (m->objective != MANAGER_SHUTDOWN) {
+                                reason = strjoina("unit ", u->id, " failed");
+                                emergency_action(m, u->failure_action, 0, u->reboot_arg, unit_failure_action_exit_status(u), reason);
+                        } else {
+                                log_info( "Tried to add Failure Action reboot job - Shutdown active" );
+                        }
```
Comment (keszybz) [inline src/core/unit.c]: Most of the other modifications in `src/core` are unexpected, and I think a localized change here is all that should be needed, or actually might be even nicer to modify emergency_action().

Basically, you want to have a condition like
```C
if (IN_SET(m->objective, MANAGER_REBOOT, MANAGER_POWEROFF, MANAGER_HALT, MANAGER_KEXEC))
    /* log a line that the emergency action is skipped because the manager is already terminating... */
else {
    /* do the old thing */
}
```

**Comment by ejb42** [comment]
@keszybz Thanks you for your reviews. I believe these changes are ready for review. 
Here is a summary:
- All of the changes for this are in emergency_action.c 
- I left a change to the order of the checks in manager_state() in manager.c.  It will report stopping before possibly reporting starting. I found a case where a failure which would cause a shutdown during the starting phase would cause manager_state() to return starting instead of stopping.
- Also fixed your comments on the test files.

File: src/test/test-honor-first-shutdown.sh [hunk 6]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+echo "Honor first shutdown test script"
+count=1
+while [ $count -eq 1 ]; do
```
Comment (keszybz) [inline src/test/test-honor-first-shutdown.sh]: What is the purpose of the $count variable? What is the purpose of the loop?

File: src/test/test-honor-first-shutdown.service [hunk 7]
```diff
@@ -0,0 +1,11 @@
+[Unit]
+Description=Honor First Shutdown feature
+After=multi-user.target
+
+[Service]
+ExecStart=/home/test-honor-first-shutdown.sh
```
Comment (keszybz) [inline src/test/test-honor-first-shutdown.service]: `/home/` ?

File: src/test/test-honor-first-shutdown.service [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+ExecStop=pkill -9 test.sh
+FailureAction=reboot
+
+[Install]
+WantedBy=multi-user.target
```
Comment (keszybz) [inline src/test/test-honor-first-shutdown.service:11]: This test is not hooked up anywhere that I can see... Is it supposed to be doing something?

File: src/core/manager.c [hunk 9]
```diff
@@ -4245,6 +4245,11 @@ ManagerState manager_state(Manager *m) {
 
         assert(m);
 
+        /* Is the special shutdown target active or queued? If so, we are in shutdown state */
+        u = manager_get_unit(m, SPECIAL_SHUTDOWN_TARGET);
+        if (u && unit_active_or_pending(u))
+                return MANAGER_STOPPING;
+
```
Comment (keszybz) [inline src/core/manager.c:4252]: Hmm, this change looks reasonable. But it needs to be a separate commit with an appropriate title and an explanation what issue it is solving and when that issue occurs.

File: src/core/emergency-action.c [hunk 10]
```diff
@@ -27,11 +39,23 @@ void emergency_action(
...
                 const char *reason) {
+        Unit *u;
 
         assert(m);
         assert(action >= 0);
         assert(action < _EMERGENCY_ACTION_MAX);
 
+        /* Is the special shutdown target active or queued? If so, we are in shutdown state */
+        u = manager_get_unit(m, SPECIAL_SHUTDOWN_TARGET);
+        if ((u && unit_active_or_pending(u))) {
```
Comment (keszybz) [inline src/core/emergency-action.c]: So... why is manager_state() not used here?

File: src/core/emergency-action.c [hunk 11]
```diff
@@ -27,11 +39,23 @@ void emergency_action(
...
+                if ( (action == EMERGENCY_ACTION_REBOOT) ||
+                     (action == EMERGENCY_ACTION_POWEROFF) ) {
+                        log_info("EmergencyAction Special Shutdown Tgt is active Skip %s",
```
Comment (keszybz) [inline src/core/emergency-action.c]: Please no abbreviations like this. Use log_unit_info with a normal sentence.

Comment (ejb42) [reply src/core/emergency-action.c]: It does not look like emergency_action is built to use log_unit_info. If I remove the call to manager_get_unit() to use manager_state() there would be no unit available to call log_unit_info(). I can do as you request but I think I would still have to call manager_get_unit() in order to use log_unit_info(). Is that the way you would like to go with this?

File: src/test/test-honor-first-shutdown.sh (hunk 6 above)
Comment (ejb42) [reply src/test/test-honor-first-shutdown.sh]: Will remove the count. The loop provides a simulated main pid that can get killed by execstop to generate the FailureAction=reboot action during a poweroff initiated by the end.service in the added tests.

File: src/test/test-honor-first-shutdown.service (hunk 7 above)
Comment (ejb42) [reply src/test/test-honor-first-shutdown.service]: Changed to / per the added test

File: src/test/test-honor-first-shutdown.service (hunk 8 above)
Comment (ejb42) [reply src/test/test-honor-first-shutdown.service:11]: Created a new test/TEST-47-HONORFIRSTSHUTDOWN/ which will be seen when we resolve the question below.

File: src/core/manager.c (hunk 9 above)
Comment (ejb42) [reply src/core/manager.c:4252]: This has been done with PR #15330

File: src/core/emergency-action.c (hunk 10 above)
Comment (ejb42) [reply src/core/emergency-action.c]: I think changing to manager_state would be the cleanest way as follows:
        /* Is the special shutdown target active or queued? If so, we are in shutdown state */
        if ((manager_state(m) == MANAGER_STOPPING) &&
            ((action == EMERGENCY_ACTION_REBOOT) ||
             (action == EMERGENCY_ACTION_POWEROFF))) {
                log_info("EmergencyAction: Shutdown is already active Skipping %s request",
                         emergency_action_table[action]);
                return;
        }
However, this change would not make it easy to use log_unit_info() per the below question.

**Comment by keszybz** [comment]
> Delete manager.c 

Hmm, what is going on here?

**Comment by ejb42** [comment]
@keszybz 
I have rebased my fork to the latest systemd and as a result I have this branch in an un-recoverable state. I believe my only recourse at this point is to create a new branch and pull request which I will do named "honor-first-shutdown-request"

**Comment by ejb42** [comment]
Zbigniew,

I apologize for the confusion. That deletion was done out of my ignorance on how to
recover my mistake on that branch. I felt my best approach would be to create a new
branch and have opened PF #15400 to continue this change.
Please let me know if this is a problem and how I should proceed.

Thank you for your patience,
-Jay
